pkg/controller/apiextensions/claim/reconciler.go

```diff
@@ -294,44 +281,7 @@ func (r *Reconciler) Reconcile(req reconcile.Request) (reconcile.Result, error)
 		record = record.WithAnnotations("composite-name", cm.GetResourceReference().Name)
 		log = log.WithValues("composite-name", cm.GetResourceReference().Name)
 
-		err := r.client.Get(ctx, meta.NamespacedNameOf(ref), cp)
-		if kerrors.IsNotFound(err) {
-
-			// Our composite was not found, but we're being deleted too. There's
-			// nothing to finalize.
-			if meta.WasDeleted(cm) {
-				// TODO(negz): Can we refactor to avoid this deletion logic that
-				// is almost identical to the meta.WasDeleted block below?
-				log = log.WithValues("deletion-timestamp", cm.GetDeletionTimestamp())
-				if err := r.claim.RemoveFinalizer(ctx, cm); err != nil {
-					// If we didn't hit this error last time we'll be requeued
-					// implicitly due to the status update. Otherwise we want to retry
-					// after a brief wait, in case this was a transient error.
-					log.Debug("Cannot remove finalizer", "error", err, "requeue-after", time.Now().Add(aShortWait))
-					record.Event(cm, event.Warning(reasonDelete, err))
-					return reconcile.Result{RequeueAfter: aShortWait}, nil
-				}
-
-				// We've successfully deleted our claim and removed our finalizer. If we
-				// assume we were the only controller that added a finalizer to this
-				// claim then it should no longer exist and thus there is no point
-				// trying to update its status.
-				log.Debug("Successfully deleted composite resource claim")
-				return reconcile.Result{Requeue: false}, nil
-			}
-
-			// If the composite resource we explicitly reference doesn't exist yet
-			// we want to retry after a brief wait, in case it is created. We
-			// must explicitly requeue because our EnqueueRequestForClaim
-			// handler can only enqueue reconciles for composite resources that
-			// have their claim reference set, so we can't expect to be queued
-			// implicitly when the composite resource we want to bind to appears.
-			log.Debug("Referenced composite resource not found", "requeue-after", time.Now().Add(aShortWait))
-			record.Event(cm, event.Warning(reasonBind, err))
-			cm.SetConditions(Waiting())
-			return reconcile.Result{RequeueAfter: aShortWait}, errors.Wrap(r.client.Status().Update(ctx, cm), errUpdateClaimStatus)
-		}
-		if err != nil {
+		if err := r.client.Get(ctx, meta.NamespacedNameOf(ref), cp); resource.IgnoreNotFound(err) != nil {
```

Comment:
Are we still guarding against the case that we have a reference but it has been deleted? If I am seeing this correctly, it looks like we will always try to get if we have a reference, and if that reference does not exist then we will requeue.

Reply:
Previously I believe we were guarding against the case in which we had a reference to an XR, but that XR didn't yet exist. i.e. The case in which an XRC wants to claim an existing XR, and the claim is created before the XR.

With this new logic if the XRC explicitly references an XR that does not yet (or anymore) exist, we'll simply continue and make a new composite resource. When Configure is called we'll overwrite the resource reference to point to the new composite resource.

I actually forget whether this change was intentional, or accidental. 😬 In retrospect I think we should probably maintain the previous behaviour; I'll update.

Reply:
I've pushed a commit that reverts to the old behaviour and also addresses the old TODO about not replicating the deletion logic. Thanks for catching this!